designs/inline-io.md

```diff
@@ -0,0 +1,207 @@
+        /// operation docs will take over that role.
+        @sensitive
+        @references([{resource: User}]) {
+            userId: String
+        }
+
+    // If there's only one trait and it's short, this compact form can be used.
+    // The references trait is the most likely trait to be used in the future,
+    // and in most cases it will be able to use this compact form.
+    output := @references([{resource: User}]) {
+        username: String
+        userId: String
+    }
+}
+
+// Inlined inputs/outputs with mixins.
+operation GetUser {
+    // Inlined io with mixins and explicit name.
+    input := GetUserRequest with BaseUser {}
+
+    // Inlined io with mixins and derived name.
+    output := with BaseUser {
+        username: String
+    }
+}
+```
+
+### ABNF
+
+```
+operation_statement =
+    "operation" ws identifier ws "{" *operation_body ws "}"
+
+operation_body =
+    operation_input / operation_output / operation_errors
+
+operation_input =
+    "input" ws (io_reference / inline_structure)
+
+operation_output =
+    "output" ws (io_reference / inline_structure)
+
+io_reference =
+    ":" ws shape_id ws
+
+inline_structure =
+    ":=" ws trait_statements [identifier] [mixins ws] "{" shape_members "}"
+
+operation_errors =
+    "errors" ws ":" ws "[" *(ws shape_id) ws "]"
+```
+
+## FAQ
+
+### Can apply be used on inlined inputs/outputs?
+
+Yes. This is only syntactic sugar, the shapes produced are normal shapes in
+every way.
+
+### Can inlined shapes be used anywhere else?
+
+No. There aren't many other places where inline shapes would make sense. Errors,
+for instance, can’t use generated names and are frequently referenced elsewhere.
+
+Consider the following simplified model that uses theoretical inlined, nested
+structure definition:
+
+```
+structure Foo {
+    bar := {
+        id: String
+    }
+}
+```
+
+On day one, the `bar` structure is only referenced in one place, so perhaps
+there’s a desire to inline it. On day two, another structure is introduced that
+references it.
+
+```
+structure Foo {
+    bar := {
+        id: String
+    }
+}
+
+structure Baz {
+    bar: bar
+}
+```
+
+Now the fact that it's inlined has become a detriment, because it's hard to go
+from looking at the definition of `Baz` to finding the definition of `bar`. This
+problem gets worse and worse the larger the model gets and the more the nested
+structure is referenced. This isn't so much a problem for operations, because
+their IO shapes are almost never refrenced elsewhere and even if they were the
+default name makes it pretty clear where to look.
+
+There is at least one other place where this may make sense: resource
+identifiers. When defining a resource, you could use this syntax to define a
+structure that contains only the resource identifiers. This could be mixed in
+to other shapes in the model. This usage, while interesting, is out of scope
+for this document.
+
+### Can the suffix be customized?
```

Comment:
Being able to customize the name of the input and output type would be something most _new_ models would never need to do or care about. However, teams migrating older models to Smithy _will_ want to maintain their existing conventions. For example, EC2 uses Request/Result I think. They wouldn't be able to use this feature, which would be a huge inconvenience for them.

I came up with a reasonable option: using a control statement. Control statements are per/file parser directives. Currently the only control statement we support is `$version`, but we can backward compatibly add other statements. We could, for example, add the following control statements:

```
$version: "2"
$operationInputSuffix: "Request"
$operationOutputSuffix: "Result"

namespace smithy.example

operation GetFoo {
    input:= {}
    output:= {}
}
```

The above would use the parser directives and create GetFooRequest and GetFooResult.

The drawback of this is that it's per/file, so it would be a minor inconvenience for service teams to remember to add these control statements. However, using validators, teams can easily catch when this is accidentally omitted (and most teams that care about maintain their existing naming schemes use tons of custom validators).